File: android/app/src/main/java/com/everittventures/salonboothmath/AdConsentManager.kt
package com.everittventures.salonboothmath

import android.app.Activity
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.google.android.gms.ads.MobileAds
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform

object AdConsentManager {
    var canRequestAds by mutableStateOf(false)
        private set

    var privacyOptionsRequired by mutableStateOf(false)
        private set

    private var mobileAdsStarted = false

    fun requestConsent(activity: Activity) {
        val consentInformation = UserMessagingPlatform.getConsentInformation(activity)
        val parameters = ConsentRequestParameters.Builder().build()

        consentInformation.requestConsentInfoUpdate(
            activity,
            parameters,
            {
                refreshState(consentInformation)
                startMobileAdsIfAllowed(activity)

                UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity) {
                    refreshState(consentInformation)
                    startMobileAdsIfAllowed(activity)
                }
            },
            {
                refreshState(consentInformation)
                startMobileAdsIfAllowed(activity)
            }
        )
    }

    fun presentPrivacyOptions(activity: Activity) {
        val consentInformation = UserMessagingPlatform.getConsentInformation(activity)
        UserMessagingPlatform.showPrivacyOptionsForm(activity) {
            refreshState(consentInformation)
            startMobileAdsIfAllowed(activity)
        }
    }

    private fun refreshState(consentInformation: ConsentInformation) {
        canRequestAds = consentInformation.canRequestAds()
        privacyOptionsRequired =
            consentInformation.privacyOptionsRequirementStatus ==
                ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED
    }

    private fun startMobileAdsIfAllowed(activity: Activity) {
        if (!canRequestAds || mobileAdsStarted) return
        mobileAdsStarted = true
        MobileAds.initialize(activity)
    }
}

File: ios/SalonBoothMath/AdConsentManager.swift
import Foundation
import GoogleMobileAds
import UserMessagingPlatform

@MainActor
final class AdConsentManager: ObservableObject {
    static let shared = AdConsentManager()

    @Published private(set) var canRequestAds = false
    @Published private(set) var privacyOptionsRequired = false

    private var mobileAdsStarted = false

    private init() {}

    func requestConsent() {
        let parameters = RequestParameters()

        ConsentInformation.shared.requestConsentInfoUpdate(with: parameters) { [weak self] _ in
            Task { @MainActor in
                guard let self else { return }

                self.refreshState()
                self.startMobileAdsIfAllowed()

                do {
                    try await ConsentForm.loadAndPresentIfRequired(from: nil)
                } catch {
                    // A previous valid consent state can still permit ads.
                }

                self.refreshState()
                self.startMobileAdsIfAllowed()
            }
        }
    }

    func presentPrivacyOptions() async {
        do {
            try await ConsentForm.presentPrivacyOptionsForm(from: nil)
        } catch {
            return
        }

        refreshState()
        startMobileAdsIfAllowed()
    }

    private func refreshState() {
        canRequestAds = ConsentInformation.shared.canRequestAds
        privacyOptionsRequired =
            ConsentInformation.shared.privacyOptionsRequirementStatus == .required
    }

    private func startMobileAdsIfAllowed() {
        guard canRequestAds, !mobileAdsStarted else { return }
        mobileAdsStarted = true
        MobileAds.shared.start()
    }
}
